### DISCUSSION TITLE
Which lemmas are destroyed on `pop`?

### DISCUSSION THREAD

**Discussion by sreeshmaheshwar** [Z3Prover/z3#7239] [Q&A]
In the following use case (using the default `smt` engine, not `sat` and no simplifiers)

```
 ; Activate solver 2, and add a bunch of assertions
(push) (assert (...)) (assert (...)) (assert (...)) (assert (...))

; Querying a single one repeatedly, surrounded by push/pop
(push) (assert (...)) (check-sat) (pop) 
(push) (assert (...)) (check-sat) (pop)
...
```

am I _actually_ solving incrementally using prior learned lemmas (I'm pretty sure I am but not concretely sure why)? This https://stackoverflow.com/a/23739636 makes me think that lemmas regarding the `assert`s at the start are maintained after the `pop` and so can be reused, but https://stackoverflow.com/a/40427658 makes me think that returning to the frozen state on `pop` discards all lemmas learnt during `(check-sat)` (which I'd guess is all of them?).

**Comment by NikolajBjorner** [comment]
All assertions are removed after pop.

**Reply by sreeshmaheshwar** [reply]
Sorry, just to clarify then @NikolajBjorner - all lemmas derived during the `(check-sat)`s are destroyed (even those relevant to the assertions that remain post-`pop`)?

**Reply by sreeshmaheshwar** [reply]
I am also wondering whether lemmas are derived on the `(assert)`s then.

**Comment by NikolajBjorner** [comment]
It is unsound to keep lemmas and assertions created under a push.
